### PR TITLE
[otbn] Some initial notes about functional coverage

### LINKED ISSUE #5959: [otbn] Produce an initial functional coverage document / framework
This will probably be folded into something more general once the OpenTitan framework is done, but get something started now to avoid being stalled.

> estimate 8

### PR DESCRIPTION
**Pull request by rswarbrick**
This text should eventually be the thing that gets signed off as
describing all the relevant coverage points. We'll either keep this
document, and connect it up to automatic coverage collection, or move
the guts of this document somewhere else when there's a better place
for it.

So far, there are a couple of general notes, plus some
specific (architectural) coverage points for the base instruction set.

This works towards #5959 (but I ended up doing things in a slightly different order from what I expected). Still probably better to mark this as fixes #5959, and then separately track groups of functional coverage that need defining.

### CODE REVIEW

File: hw/ip/otbn/doc/fcov.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+Because OTBN doesn't have a complicated multi-stage pipeline or any real exception handling, we don't track much temporal information (such as sequences of instructions).
+
+As well as instruction-specific coverage points detailed below, we include a requirement that each instruction is executed at least once.
+
+For any instruction with one or more immediate fields, we require "toggle coverage" for those fields.
+That is, we expect to see execution with each bit of each immediate field being zero and one.
+We also expect to see each field with values `'0` and `'1` (all zeros and all ones).
+If the field is treated as a signed number, we also expect to see it with the extremal values for its range (just the MSB set, for the most negative value; all but the MSB set, for the most positive value).
+
+For any instruction that reads from or writes to a GPR register, we expect to see that operand equal to `x0`, `x1` and an arbitrary register in the range `x2 .. x15`.
```
Comment (GregAC) [inline hw/ip/otbn/doc/fcov.md:49]: `x2 .. x31`?

File: hw/ip/otbn/doc/fcov.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+### ECALL
+
+No special coverage points for this instruction.
+
+### LOOP
+
+- Loop with a zero iteration count (causing an error)
+- Loop with a count of `'1` (the maximal value)
+- Loop when the match address would be above the top of memory.
```
Comment (GregAC) [inline hw/ip/otbn/doc/fcov.md:230]: By match address do you mean loop end address?

File: hw/ip/otbn/doc/fcov.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+- Loop when the loop stack is full, causing an overflow.
+- Loop at the end of a loop.
+- Duplicate loop match address, matching top of stack
+- Duplicate loop match address, further down stack
```
Comment (GregAC) [inline hw/ip/otbn/doc/fcov.md:234]: Assuming match address is end address have we architecturally defined what's happening here and do we want to? I I don't think the current ISA docs don't explicitly say what should happen though it should be easy enough for the RTL and ISS to match their behaviour.

File: hw/ip/otbn/doc/fcov.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+## Call stack
+
+The [call stack]({{< relref ".#call-stack" >}}) is exposed as a special register behind `x1`.
+It has a bounded depth of 8 elements.
+We expect to see the following events:
+
+- Push to the call stack
+- Pop from the call stack
+- Push and pop from the call stack on a single instruction
```
Comment (GregAC) [inline hw/ip/otbn/doc/fcov.md:28]: Push/Pop to/from full/empty call stack. Including the simultaneous push and pop when full (which should be allowed).

**Review by GregAC — COMMENTED** [review]
Looks like a good start, a few minor comments

File: hw/ip/otbn/doc/fcov.md (hunk 4 above)
Comment (rswarbrick) [reply hw/ip/otbn/doc/fcov.md:28]: I think this is covered by the next sentence?

File: hw/ip/otbn/doc/fcov.md (hunk 3 above)
Comment (rswarbrick) [reply hw/ip/otbn/doc/fcov.md:234]: I assume we always just compare with the top of stack (so the RTL and ISS should probably just match). We don't special-case these cases in the spec, and I don't think there's any need to. However, I've just looked at our docs and we never actually say what LOOP and LOOPI do! (Oops...) I'll file a follow-up bug to remind us to do so.

File: hw/ip/otbn/doc/fcov.md (hunk 4 above)
Comment (GregAC) [reply hw/ip/otbn/doc/fcov.md:28]: Ah yes, somehow missed that sentence.

Have thought of another one, double call stack access (e.g. add x2, x1, x1). Will only produce a single pop so both input operands get the same value.

Comment (rswarbrick) [reply hw/ip/otbn/doc/fcov.md:28]: Ooh, good point! Added.

File: hw/ip/otbn/doc/fcov.md [hunk 5]
```diff
@@ -0,0 +1,239 @@
+---
+title: OTBN functional coverage plan
```
Comment (sriyerg) [inline hw/ip/otbn/doc/fcov.md:2]: The DV doc is precisely the place to capture this. I agree with @matutem - there is no need to put this in a separate document.

Comment (rswarbrick) [reply hw/ip/otbn/doc/fcov.md]: It's going to get awfully long: I suspect we'll change our mind about this at some point. But ok.

**Review by GregAC — APPROVED** [review]

**Pull request merged**